code/modules/photography/camera/camera.dm

```diff
@@ -0,0 +1,220 @@
+		return FALSE
+	return TRUE
+
+/obj/item/camera/afterattack(atom/target, mob/user, flag)
+	if (disk)
+		if(ismob(target))
+			if (disk.record)
+				QDEL_NULL(disk.record)
+
+			disk.record = new
+			var/mob/M = target
+			disk.record.caller_name = M.name
+			disk.record.set_caller_image(M)
+		else
+			to_chat(user, "<span class='warning'>Invalid holodisk target.</span>")
+			return
+
+	if(!can_target(target, user, flag))
+		return
+
+	on = FALSE
+	addtimer(CALLBACK(src, .proc/cooldown), cooldown)
+	icon_state = state_off
+
+	INVOKE_ASYNC(src, .proc/captureimage, target, user, flag, picture_size_x, picture_size_y)
+
+
+/obj/item/camera/proc/cooldown()
+	UNTIL(!blending)
+	icon_state = state_on
+	on = TRUE
+
+/obj/item/camera/proc/show_picture(mob/user, datum/picture/selection)
+	var/obj/item/photo/P = new(src, selection)
+	P.show(user)
+	to_chat(user, P.desc)
+	qdel(P)
+
+/obj/item/camera/proc/captureimage(atom/target, mob/user, flag, size_x = 1, size_y = 1)
+	blending = TRUE
+	var/turf/target_turf = get_turf(target)
+	if(!isturf(target_turf))
+		blending = FALSE
+		return FALSE
+	size_x = CLAMP(size_x, 0, CAMERA_PICTURE_SIZE_HARD_LIMIT)
+	size_y = CLAMP(size_y, 0, CAMERA_PICTURE_SIZE_HARD_LIMIT)
+	var/list/desc = list()
+	var/ai_user = isAI(user)
+	var/list/seen
+	var/list/viewlist = (user && user.client)? getviewsize(user.client.view) : getviewsize(world.view)
+	var/viewr = max(viewlist[1], viewlist[2]) + max(size_x, size_y)
+	var/viewc = user.client? user.client.eye : target
+	seen = get_hear(viewr, viewc)
+	var/list/turfs = list()
+	var/list/mobs = list()
+	var/blueprints = FALSE
+	var/clone_area = SSmapping.RequestBlockReservation(size_x * 2 + 1, size_y * 2 + 1)
+	for(var/turf/T in block(locate(target_turf.x - size_x, target_turf.y - size_y, target_turf.z), locate(target_turf.x + size_x, target_turf.y + size_y, target_turf.z)))
+		if((ai_user && GLOB.cameranet.checkTurfVis(T)) || T in seen)
+			turfs += T
+			for(var/mob/M in T)
+				mobs += M
+			if(locate(/obj/item/areaeditor/blueprints) in T)
+				blueprints = TRUE
+	for(var/i in mobs)
+		desc += camera_get_mobdesc(i)
+
+	var/psize_x = (size_x * 2 + 1) * world.icon_size
+	var/psize_y = (size_y * 2 + 1) * world.icon_size
+	var/get_icon = camera_get_icon(turfs, target_turf, psize_x, psize_y, clone_area, size_x, size_y, (size_x * 2 + 1), (size_y * 2 + 1))
+	qdel(clone_area)
+	var/icon/temp = icon('icons/effects/96x96.dmi',"")
+	temp.Blend("#000", ICON_OVERLAY)
+	temp.Scale(psize_x, psize_y)
+	temp.Blend(get_icon, ICON_OVERLAY)
+
+	var/datum/picture/P = new("picture", desc.Join(""), temp, null, psize_x, psize_y, blueprints)
+	after_picture(user, P, flag)
+	blending = FALSE
+
+/obj/item/camera/proc/camera_get_mobdesc(mob/M)
+	var/list/mob_details = list()
+	if(M.invisibility)
+		if(see_ghosts && isobserver(M))
+			mob_details += "You can also see a g-g-g-g-ghooooost! "
+		else
+			return mob_details
+	var/list/holding = list()
+	if(isliving(M))
+		var/mob/living/L = M

```

Comment:
You could probably abstract these into some /get_photo_description for both ghosts and other mobs.